crates/pgt_workspace/src/workspace/server/change.rs

```diff
@@ -63,11 +63,33 @@ impl Document {
         // very much not guaranteed to result in correct ranges
         self.diagnostics.clear();
 
-        let changes = change
-            .changes
-            .iter()
-            .flat_map(|c| self.apply_change(c))
-            .collect();
+        // when we recieive more than one change, we need to push back the changes based on the
+        // total range of the previous ones. This is because the ranges are always related to the original state.
+        let mut changes = Vec::new();
+
+        let mut offset: i64 = 0;
+
+        for change in &change.changes {
+            let adjusted_change = if offset != 0 && change.range.is_some() {
+                &ChangeParams {
+                    text: change.text.clone(),
+                    range: change.range.map(|range| {
+                        let start = u32::from(range.start());
+                        let end = u32::from(range.end());
+                        TextRange::new(
+                            TextSize::from((start as i64 + offset).try_into().unwrap_or(0)),
+                            TextSize::from((end as i64 + offset).try_into().unwrap_or(0)),
```

Comment:
nit: isn't this `range.checked_add(offset)`?

Reply:
The problem with checked_add is that it expects a TextSize which is always > 0.